### ISSUE TITLE
Shorter links for ModelAdmin items

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-admin#1075]
Right now an item on a ModelAdmin has the URL of:
`/admin/(modeladmin)/(tab)/EditForm/field/(tab)/item/214/edit`

This would be usefully contracted to without losing ambiguity:
`/admin/(modeladmin)/(tab)/214`

 * Numeric values after the tab can be interpreted as referring to the item request on the ModelAdmin's default GridField field
 * The default action (edit) can be referenced without redirection

Changing the URLs in this way could be disruptive and so I would introduce it as an optionally enabled feature of ModelAdmin in 4.x.

**Comment by sminnee** [comment]
@chillu make sense to you?

**Comment by chillu** [comment]
Yeah that makes a lot of sense. Context for others, this builds on a pending PR in https://github.com/silverstripe/silverstripe-admin/pull/1072. 

I wonder if it's better to hide the complexity away from GridField and handle this within ModelAdmin only. Maybe we can make `GridFieldDetailForm->handleItem()` a bit smarter, and tell those particular gridfields to behave differently. Or maybe that's what you mean by "monkeypatching".

My only concern is that we're breaking URL symmetry with `CMSMain` here: `admin/pages/edit/show/4`. It's admittedly always been confusing to have `edit` (referring to the "edit controller") followed by `show` (referring to the action within that controller). How about `/admin/(modeladmin)/(tab)/edit/214`? That would also leave the door open to a separate `view/` route later. View vs. edit isn't only decided on whether you have permissions, it's sometimes just better to look through the readonly view (e.g. with formatted markdown, without a clunky HTML editor component).

**Comment by sminnee** [comment]
> How about /admin/(modeladmin)/(tab)/edit/214?

Sounds fine. If someone adds extra actions to the GridField without also modifying ModelAdmin you would end up with something like `/admin/(modeladmin)/(tab)/edit/214/view`, which is a bit weird but relatively benign, and can be solved with further ModelAdmin improvements. It might make sense to customise both the view & edit URLs in the initial implementation just to validate that it's possible.

> I wonder if it's better to hide the complexity away from GridField and handle this within ModelAdmin only. 

Something like this makes sense to me

```php
$gf->getConfig()->getComponentByType(GridFieldDetailForms::class)
->setItemLinkCallback(function ($record, $action) use ($modelAdmin) {
  return Controller::join_links($modelAdmin->Link(), $modelAdmin->tabName, 'edit', $record->ID);
})
```

So we just need to define `GridFieldDetailForms::setItemLinkCallback()`.

**Comment by GuySartorelli** [comment]
I don't think this would be valuable anymore - the changes made as part of https://github.com/silverstripe/silverstripe-admin/issues/1354 trivialise getting a canonical edit URL for any given record, and since this isn't the front-end the value of having short URLs people can remember is basically nil.
